test/math_libraries/convolutions.py

```diff
@@ -0,0 +1,120 @@
+    dtypes, dtypesIfCUDA, skipCUDAIfRocm, onlyOnCPUAndCUDA
+from torch.testing._internal.common_nn import NNTestCase
+from torch.testing._internal.common_utils import run_tests
+import os
+import json
+import warnings
+
+class TestConvExt(NNTestCase):
+    def _subtensor(self, tensor, dim, group, g):
+        if tensor is None:
+            return None
+        group_size = int(tensor.size(dim) / group)
+        return tensor.narrow(dim, group_size * g, group_size).contiguous()
+
+    def _thnn_conv(self, input, weight, k, bias, s, p, d):
+        if d[0] > 1 or d[1] > 1:
+            return torch._C._nn.slow_conv_dilated2d(input, weight, k, bias, s, p, d)
+        else:
+            return torch._C._nn.thnn_conv2d(input, weight, k, bias, s, p)
+
+    def _thnn_conv_group(self, input, weight, k, bias, s, p, d, group):
+        if group == 1:
+            return self._thnn_conv(input, weight, k, bias, s, p, d)
+        else:
+            outputs = []
+            for g in range(group):
+                input_g = self._subtensor(input, 1, group, g)
+                weight_g = self._subtensor(weight, 0, group, g)
+                bias_g = self._subtensor(bias, 0, group, g)
+                outputs.append(self._thnn_conv(input_g, weight_g, k, bias_g, s, p, d))
+            return torch.cat(outputs, 1)
+
+    def _collect_cases(self):
+        dir_name = 'convolutions_cases'
+        cur_dir = os.path.split(os.path.realpath(__file__))[0]
+        case_files = ['googlenet_v3', 
+                      'maskrcnn_p1', 
+                      'mobilenet', 
+                      'resnet_50']
+        for index, case_file in enumerate(case_files):
+            case_files[index] = '{}/{}/shapes_{}.json'.format(cur_dir, dir_name, case_file)
+        total_cases = list()
+        for file_name in case_files:
+            with open(file_name) as f:
+                model_cases = json.loads(f.read())
+            total_cases += model_cases
+        return total_cases
+
+    # Set multiple situations to improve convolution2d test coverage
+    @onlyOnCPUAndCUDA
+    @skipCUDAIfRocm
+    @dtypesIfCUDA(torch.half, torch.float, torch.double)
+    @dtypes(torch.float)
+    def test_conv2d_ext(self, device, dtype):
+        # this list for save the cases which dh,dw == 0
+        # and raise error in the end
+        invalid_cases = list()
+
+        total_cases = self._collect_cases()
+        for case in total_cases:
+            case_name = case['case_name']
+            bs = case['mb']
+            group = case['g']
+            ic, ih, iw = case['ic'], case['ih'], case['iw']
+            oc = case['oc']
+            kh, kw = case['kh'], case['kw']
+            sh, sw = case['sh'], case['sw']
+            ph, pw = case['ph'], case['pw']
+            dh, dw = case['dh'], case['dw']
+            has_bias = case['bias']
+            if dh == 0 or dw == 0:
+                invalid_cases.append(case_name)
```

Comment:
Much easier would be to `self.Fail` here and avoid `invalid_cases` construct. This will prevent json-s to be submitted with wrong inputs in the first place.